lib/experimental/html/class-wp-html-character-reference-transcoder.php

```diff
@@ -0,0 +1,1149 @@
+	 * @return string String with character references replaced by their corresponding text.
+	 */
+	public static function decode_utf8( $context, $input ) {
+		$at     = 0;
+		$buffer = '';
+		$budget = 1000;
+		$end    = strlen( $input );
+
+		while ( $at < $end && $budget-- > 0 ) {
+			$next = strpos( $input, '&', $at );
+			/*
+			 * We have to have at least as many successive characters as
+			 * can be used to find the character reference group. The
+			 * shortest named character reference is three characters, so
+			 * we need at least this many.
+			 */
+			if ( false === $next || $next + 3 > $end ) {
+				break;
+			}
+
+			// Grab content up until this point.
+			$buffer .= substr( $input, $at, $next - $at );
+
+			// Skip the `&` for continued processing.
+			$at = $next + 1;
+
+			// Handle decimal and hex numeric inputs
+			if ( '#' === $input[ $at ] ) {
+				if ( 'x' === $input[ $at + 1 ] || 'X' === $input[ $at + 1 ] ) {
+					$numeric_base   = 16;
+					$numeric_digits = '0123456789abcdefABCDEF';
+					$max_digits     = 6; // &#x10FFFF;
+					$at += 2;
+				} else {
+					$numeric_base   = 10;
+					$numeric_digits = '0123456789';
+					$max_digits     = 7; // &#1114111;
+					$at += 1;
+				}
+
+				// Leading zeros are interpreted as zero values; skip them.
+				$at += strspn( $input, '0', $at );
+				if ( $at === $end ) {
+					$buffer .= substr( $input, $next, $at - $next );
+					break;
+				}
+
+				// Max legitimate character reference is to U+10FFFF.
+				$digit_count = strspn( $input, $numeric_digits, $at );
+				if ( 0 === $digit_count || $digit_count > $max_digits ) {
+					$at += $digit_count;
+					$buffer .= substr( $input, $next, $at - $next );
+					continue;
+				}
+				$digits     = substr( $input, $at, $digit_count );
+				$code_point = intval( $digits, $numeric_base );
+
+				/*
+				 * While HTML specifies that we replace invalid references like these
+				 * with the replacement character U+FFFD, we're going to leave it in
+				 * so we can preserve the input as best we can. The browser will still
+				 * replace it eventually, but until render we don't want to inject
+				 * these replacement characters into the data stream.
+				 */
+				if (
+					// Null character.
+					0 === $code_point ||
+
+					// Outside Unicode range.
+					$code_point > 0x10FFFF ||
+
+					// Surrogate.
+					( $code_point >= 0xD800 && $code_point <= 0xDFFF ) ||
+
+					// Noncharacters.
+					( $code_point >= 0xFDD0 && $code_point <= 0xFDEF ) ||
+					( 0xFFFE === ( $code_point & 0xFFFE ) ) ||
+
+					// 0x0D or non-ASCII-whitespace control
+					0x0D === $code_point ||
+					(
+						$code_point >= 0 &&
+						$code_point <= 0x1F &&
+						0x9 !== $code_point &&
+						0xA !== $code_point &&
+						0xC !== $code_point &&
+						0xD !== $code_point
+					)
+				) {
+					$at += $digit_count;
+					$buffer .= substr( $input, $next, $at - $next );
+					continue;
```

Comment:
Nit: just `continue;` should suffice here as well – the next loop iteration will skip over the digits anyway.